vyper/venom/analysis/loop_detection.py

```diff
@@ -0,0 +1,45 @@
+from vyper.venom.analysis.analysis import IRAnalysis
+from vyper.venom.analysis.cfg import CFGAnalysis
+from vyper.venom.basicblock import IRBasicBlock
+
+
+class LoopDetection(IRAnalysis):
+    """
+    Detects loops and computes basic blocks sets
+    which comprised these loops
+    """
+
+    # key = start of the loop (last bb not in the loop), value all the block that loop contains
+    loops: dict[IRBasicBlock, list[IRBasicBlock]]
+
+    def analyze(self):
+        self.analyses_cache.request_analysis(CFGAnalysis)
+        self.loops: dict[IRBasicBlock, list[IRBasicBlock]] = dict()
+        done: dict[IRBasicBlock, bool] = dict(
```

Comment:
We usually use a `set` for this purpose, and you will see that we also tent to use our own `OrderedSet` implementation even when really not _visibly_ necessary, to avoid ending up with some weird non-deterministic code generation.